components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/AmrDecisionEvaluator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.handler.sequence.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.AuthenticationDecisionEvaluator;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthHistory;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.common.model.ServiceProvider;
+import org.wso2.carbon.identity.application.common.model.graph.DecisionNode;
+import org.wso2.carbon.identity.application.common.model.graph.Link;
+
+import java.util.List;
+
+/**
+ * Performs the authentication decision based on ACR in the context.
+ */
+public class AmrDecisionEvaluator implements AuthenticationDecisionEvaluator {
+
+    private static final Log log = LogFactory.getLog(AmrDecisionEvaluator.class);
+
+    @Override
+    public String evaluate(AuthenticationContext context, ServiceProvider serviceProvider, DecisionNode config) {
+        List<String> amrRequested = context.getAmrRequested();
+        if (amrRequested == null || amrRequested.isEmpty()) {
+            if (log.isDebugEnabled()) {
+                log.debug("AMR values from context is empty. Selecting the default outcome (i.e. null).");
+            }
+            return null;
+        }
+        List<AuthHistory> histories = context.getAuthenticationStepHistory();
+        List<Link> links = config.getLinks();
+        Link selected = links.stream().filter(l -> l.getExpression().equals(amrRequested.stream()
+                .filter(m -> !histories.stream().filter(h -> h.getAuthenticatorName().equals(m)).findAny()
+                        .isPresent()))).findAny().orElse(null);
```

Comment:
Instead of calling list.stream().findAny().isPresent(), it is more efficient to use list.stream().anyMatch(). It is also the correct idiom for expressing you intention.

https://stackoverflow.com/a/39721919/388714